src/build/app/mod.rs

```diff
@@ -2121,12 +2142,47 @@ impl<'help> App<'help> {
     ///     .unwrap_or_else(|e| e.exit());
     /// ```
     /// [`App::try_get_matches_from`]: App::try_get_matches_from()
+    ///
+    /// ### Panics
+    ///
+    /// If both [`AppSettings::Multicall`]
+    /// and [`AppSettings::NoBinaryName`] are used.
     pub fn try_get_matches_from_mut<I, T>(&mut self, itr: I) -> ClapResult<ArgMatches>
     where
         I: IntoIterator<Item = T>,
         T: Into<OsString> + Clone,
     {
+        if self.settings.is_set(AppSettings::NoBinaryName)
+            && self.settings.is_set(AppSettings::Multicall)
+        {
+            panic!("NoBinaryName can't be used with Multicall");
+        }
+
         let mut it = Input::from(itr.into_iter());
+
+        if self.settings.is_set(AppSettings::Multicall) {
+            if let Some((argv0, _)) = it.next() {
+                let argv0 = Path::new(&argv0);
+                if let Some(command) = argv0.file_name().and_then(|f| f.to_str()) {
+                    match self
+                        .subcommands
+                        .iter_mut()
+                        .find(|subcommand| subcommand.aliases_to(command))
+                    {
+                        Some(subcommand) => {
+                            self.name.clear();
+                            self.bin_name = None;
+                            it.insert(&[subcommand.get_name()]);
+                            return self._do_parse(&mut it);
+                        }
+                        _ => {
+                            self.bin_name.get_or_insert_with(|| command.to_owned());
+                            return self._do_parse(&mut it);
+                        }
+                    };
```

Comment:
From #1120

> The basic premise would be argv[0] would be matched in src/app/mod.rs:get_matches_from_safe_borrow first. If it matches self.p.meta.name (i.e. what's given to App::new), everything functions exactly like it already does in clap (i.e. all the "applets" are just subcommands).

> However, if it doesn't match, you could check that it matches a subcommand name and set self.p.meta.bin_name and self.p.meta.name to a blank string and simply continue parsing like normal because clap will simply treat it as a subcommand. The reason you set self.p.meta.bin_name is to that all usage strings and error messages say applet [options] and not q applet [options].

This PR seems to only check if its a subcommand, and if it isn't, it assumes its the regular bin. It skips the `self.p.meta.name` check in the first quoted paragraph.

The one thing @kbnapp left off of #1120 was what to do if the bin doesn't match a subcommand.  If its to fall through, then this implementation still has the same runtime behavior as what was suggested with different performance characteristics with which being a winner only clear with use cases and benchmarks (`self.p.meta.name` helps when using the base command regularly, skipping it helps when using applets regularly)

On the other hand, if we have the `self.p.meta.name` check, we could make the subcommand logic unconditional (no search) which would allow for supporting external subcommands.  I fully admit this is probably strange.  I've generally found erring on the side of composability of features makes the core easier to read, expand, and for users to create solutions you never thought possible (in a positive way).

Reply:
> let m = app.clone().get_matches_from(&["a.out", "--install"]);

This is taken from the tests and, I assume, the reason we only march subcommands when the name matches one.

Reply:
Oh, its to give a subcommand primacy over the bin

> This handles the feature from the linked issue of hostname-like behaviour where applets shouldn't be callable from the main command, by having a subcommand with the same name as the main command i.e. hostname has a hostname subcommand and being called with the executable name hostname runs the hostname subcommand name and runs the dnsdomainname subcommand if linked as dnsdomainname

See also https://github.com/clap-rs/clap/issues/1120#issuecomment-349376005

Reply:
> This doesn't handle the busybox feature of nicely reporting that an applet isn't found when a subcommand doesn't match.

So erroring if it doesn't match either is another idea for us to consider here

Reply:
To summarize our options

Should base command or applets get precedence?  This PR favors applets, mimicking the `hostname` command

What should we do if its not the base command or applet?
- Still fallback, like base command (what this PR currently does)
- Error here
- Treat all non-base commands as subcommands, leaving error handling to our subcommands and allowing all other subcommand features (including external subcommands)
  - This seems simpler than doing a special invalid subcommand check in here (the Error option)

I lean towards
- Give applets preference
- Treat all non-base commands as subcommands

@kbknapp since you suggested the original direction, I'd be interested in your thoughts

Reply:
I've done the initial read-through of this PR but I need to do some testing first. Unfortunately I don't have time to knock that out tonight, so I'll attack that tomorrow and respond here with my results. Just wanted to let everyone know I saw the ping and will be responding :smile: